### ISSUE TITLE
Waypoint editable graph represention mode

### ISSUE THREAD

**Issue by caryoscelus** [synfig/synfig#267]
Alternative timetrack dock/mode, which would present parameter values as a graph (as graphs dock currently does) *and* allow to edit waypoints, including their parameters (e.g. curve handles for curve interpolation)

**Comment by morevnaproject** [comment]
I think it's best to implement this in current Graph Panel.

**Comment by caryoscelus** [comment]
Shouldn't matter from user perspective and from developer perspective i have no idea which would be easier - haven't looked much at the code of either docks.

**Comment by morevnaproject** [comment]
We need that in Graph Panel. This is highly requested feature from many artists.

**Comment by Jose-Moreno** [comment]
@morevnaproject Implementing this would be terrific! I honestly tried to use and understand the TCB vs. other interpolations, but it's just not the same. I have an After Effects background and the keyframe interpolation, or what you call waypoints, always behaves in ways that the artist can predict, however for certain types of animation you are practically required to use (bezier) curves; they are key to create complicated patterns in motion graphics.

Since I've noticed more activity lately in the repo, I'll gladly test the latest dev builds and submit every detailed bug report I can, I have a track record with testing FOSS. Also I want to submit feature requests that myself and other professional colleagues of mine have made when trying out the software. 

Hopefully we can finally see Synfig getting more support and market share in the open-source graphics ecosystem :slightly_smiling_face:

**Comment by BobSynfig** [comment]
Indeed it would be nice but it would surely introduce more bugs, especially with multi-values waypoints (like colors).
Personally I would prefer to stabilze and solve crashes first :P

**Comment by morevnaproject** [comment]
@Jose-Moreno Yes, that feature is a long-awaited by all us, and (as far as I know) it is a major roadblock, which prevents professional animators form using Synfig. This is why I pinned this issue  to bring more attention to it. And it is in our roadmap - https://synfig-docs-dev.readthedocs.io/en/latest/common/roadmap.html

> Since I've noticed more activity lately in the repo, I'll gladly test the latest dev builds and submit every detailed bug report I can, I have a track record with testing FOSS. Also I want to submit feature requests that myself and other professional colleagues of mine have made when trying out the software.

Thank you, we will appreciate your help here for sure! ^__^

**Comment by morevnaproject** [comment]
I think we can split this task in several steps.

**Step 0:** here is how Graphs Panel look at the moment:

In UI context:
![screenshot_003](https://user-images.githubusercontent.com/332868/53815801-ee770d00-3f94-11e9-904b-07a328644731.png)

Separate screenshot:
![screenshot_004](https://user-images.githubusercontent.com/332868/53815836-09498180-3f95-11e9-9def-ef8b841cd1b8.png)

**Step 1:** Add markers, which corresponding to Waypoints

![screenshot_004](https://user-images.githubusercontent.com/332868/53815970-50377700-3f95-11e9-86e4-04bce2bbce63.png)

**Step 2:** Allow user to change values (curves) by dragging those markers (make markers draggable using mouse).

**Step 3:** Implementing control points for TCB Waypoints

TCB Waypoints  are shown with green dots. If you right-click on TCB Waypoint (1) and choose "Edit" then you will see a dialog with parameters, which allows tweak shape of the curve (2).

![screenshot_005](https://user-images.githubusercontent.com/332868/53817943-c938cd80-3f99-11e9-9434-4290c221e0c2.png)

Those parameters are kinda clumsy and hard to understand. So, I think it's okay to create handles that allow to change "Bias" parameter only.

![screenshot_006](https://user-images.githubusercontent.com/332868/53818354-8297a300-3f9a-11e9-8450-79eff34ff6df.png)

(On the illustration: dragging handle 1 will change  Bias parameter).


**Step 4:** Implementing Bezier Waypoints

Since TCB Waypoints are not very flexible to define curve shape, we need to introduce bew type of Waypoint, which is  defined by Bezier curve.

This is how editable Bezier Curves work in OpenToonz -
https://youtu.be/VYbP0H3D0bE

**Step 5:** Make Bezier Waypoints editable using mouse.

Add handles (as in step 3) and make them draggable with mouse.
As shown in video about OpenToonz above.

---

This is it, I hope I didn't forgot anything. ^__^

**Comment by maitreyi0505** [comment]
I would like to work on this project as a part of GSoC 2019, it would be great of I could get some guidance on ow to proceed!!

**Comment by morevnaproject** [comment]
@maitreyi0505 Great! Please consider that there is already some competition among students around “Editable animation curves in Graphs Panel” project. If you do not mind that, then I suggest to start by implementing this feature as a proof of your eligibility - https://github.com/synfig/synfig/issues/783

**Comment by morevnaproject** [comment]
> are the TCB waypoints to be implemented ,or just the Beizer ones??

TCB waypoints are already implemented. We will need Bezier waypoints implemented as part of this project.

**Comment by morevnaproject** [comment]
@rodolforg is working on this via #982.

![screenshot_001](https://user-images.githubusercontent.com/332868/67690463-8667b200-f9cf-11e9-93f9-ca6391916700.png)

**Comment by yoSachinkr** [comment]
I think Synfig could benefit from splitting curves into two graphs: a Value graph and a Speed graph. The Value graph would show the actual values of a property—like position, scale, or opacity—over time, making it easier to see how these values change throughout the animation. The Speed graph would display how quickly a property changes over time, helping to control the pacing and smoothness of motion in animations.

**Comment by rodolforg** [comment]
> The Speed graph would display how quickly a property changes over time, helping to control the pacing and smoothness of motion in animations.

@yoSachinkr  Do you mean a derivative curve?

**Comment by yoSachinkr** [comment]
> Do you mean a derivative curve?

Yes, the derivative curve.
